Mathlib/CategoryTheory/Functor/TwoSquare.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2025 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou, Jakob von Raumer
+-/
+import Mathlib.CategoryTheory.Whiskering
+import Mathlib.Tactic.CategoryTheory.Slice
+
+/-!
+# 2-squares of functors
+
+Given four functors `T`, `L`, `R` and `B`, a 2-square `TwoSquare T L R B` consists of
+a natural transformation `w : T ⋙ R ⟶ L ⋙ B`:
+```
+     T
+  C₁ ⥤ C₂
+L |     | R
+  v     v
+  C₃ ⥤ C₄
+     B
+```
+
+We define operations to paste such squares horizontally and vertically and prove the interchange
+law of those two operations.
+
+## TODO
+
+Generalize all of this to double categories.
+
+-/
+
+universe v₁ v₂ v₃ v₄ v₅ v₆ v₇ v₈ v₉ u₁ u₂ u₃ u₄ u₅ u₆ u₇ u₈ u₉
+
+namespace CategoryTheory
+
+open Category
+
+variable {C₁ : Type u₁} {C₂ : Type u₂} {C₃ : Type u₃} {C₄ : Type u₄}
+  [Category.{v₁} C₁] [Category.{v₂} C₂] [Category.{v₃} C₃] [Category.{v₄} C₄]
+  (T : C₁ ⥤ C₂) (L : C₁ ⥤ C₃) (R : C₂ ⥤ C₄) (B : C₃ ⥤ C₄)
+
+/-- A `2`-square consists of a natural transformation `T ⋙ R ⟶ L ⋙ B`
+involving fours functors `T`, `L`, `R`, `B` that are on the
+top/left/right/bottom sides of a square of categories. -/
+def TwoSquare := T ⋙ R ⟶ L ⋙ B
+
+namespace TwoSquare
+
+/-- Constructor for `TwoSquare`. -/
+abbrev mk (α : T ⋙ R ⟶ L ⋙ B) : TwoSquare T L R B := α
+
+variable {T} {L} {R} {B} in
+/-- The natural transfomration associated to a 2-square. -/
+abbrev out (w : TwoSquare T L R B) : T ⋙ R ⟶ L ⋙ B := w
+
+/-- The type of 2-squares on functors `T`, `L`, `R`, and `B` is trivially equivalent to
+the type of natural transformations `T ⋙ R ⟶ L ⋙ B`. -/
+@[simps]
+def equivNatTrans : TwoSquare T L R B ≃ (T ⋙ R ⟶ L ⋙ B) where
+  toFun := out
+  invFun := mk T L R B
+  left_inv := fun _ => rfl
+  right_inv := fun _ => rfl
+
+variable {T L R B}
+
+@[ext]
+lemma ext (w w' : TwoSquare T L R B) (h : ∀ (X : C₁), w.app X = w'.app X) : w = w' :=
+  NatTrans.ext (funext h)
+
+/-- The hoizontal identity 2-square. -/
+@[simps!]
+def hId (L : C₁ ⥤ C₃) : TwoSquare (𝟭 _) L L (𝟭 _) :=
+  𝟙 _
+
+/-- Notation for the horizontal identity 2-square. -/
+scoped notation "𝟙ₕ" => hId  -- type as \b1\_h
+
+/-- The vertical identity 2-square. -/
+@[simps!]
+def vId (T : C₁ ⥤ C₂) : TwoSquare T (𝟭 _) (𝟭 _) T :=
+  𝟙 _
+
+/-- Notation for the vertical identity 2-square. -/
+scoped notation "𝟙ᵥ" => vId  -- type as \b1\_v
+
+/-- Whiskering a 2-square with a natural transformation at the top. -/
+@[simps!]
+protected def whiskerTop {T' : C₁ ⥤ C₂} (α : T ⟶ T') (w : TwoSquare T' L R B) : TwoSquare T L R B :=
```

Comment:
In order to allow dot notation, it would be nice to put `w` as the first parameter.

Reply:
My first thought was to keep have the same change of order as `CategoryTheory.whiskerRight`, but you're right that for `CategoryTheory.whiskerRight` we don't have dot notation anyway